server/test/lib/scheduler/UserSchedule.spec.ts

```diff
@@ -1,86 +1,217 @@
+        expect(this.userSchedule.hasUserWorkedOnWorkspace(workspaces.get("2-1"))).to.equal(true);
+      });
+    });
+  });
+
+  describe("isUserCurrentlyWorkingOnWorkspace method", function() {
+    context("with a user that has not been assigned to any workspace", function() {
+      it("returns false", function() {
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("1"))).to.equal(false);
+      });
+    });
+
+    context("with a user that has been assigned to one workspace", function() {
+      it("returns true for assigned workspace", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("1"));
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("1"))).to.equal(true);
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("1-1"))).to.equal(false);
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("2"))).to.equal(false);
+      });
+    });
+
+    context("with a user that has been assigned to two workspaces", function() {
+      it("returns true for most recent, false for less recent", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("2"));
+        await this.userSchedule.assignWorkspace(workspaces.get("2-1"));
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("2"))).to.equal(false);
+        expect(this.userSchedule.isUserCurrentlyWorkingOnWorkspace(workspaces.get("2-1"))).to.equal(true);
+      });
+    });
+  });
+
+  describe("getLastWorkedOnWorkspace method", function() {
+    context("with a user that has not been assigned to workspace", function() {
+      it("returns undefined", function() {
+        expect(this.userSchedule.getLastWorkedOnWorkspace()).to.equal(undefined);
+      });
+    });
+
+    context("with a user that has been assigned to a workspace once", function() {
+      it("returns that workspace", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("1"));
+        expect(this.userSchedule.getLastWorkedOnWorkspace()).to.equal(workspaces.get("1"));
+      });
+    });
+
+    context("with a user that has been assigned to two workspaces", function() {
+      it("returns the most recently assigned workspace", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("1"));
+        await this.userSchedule.assignWorkspace(workspaces.get("2"));
+        expect(this.userSchedule.getLastWorkedOnWorkspace()).to.equal(workspaces.get("2"));
+      });
+    });
+  });
+
+  describe("isActiveInLastWorkspace method", function() {
+    context("with a user that has not been assigned to workspace", function() {
+      it("returns undefined", function() {
+        expect(this.userSchedule.getLastWorkedOnWorkspace()).to.equal(undefined);
+      });
+    });
+
+    context("with a user that has been assigned a workspace and no time has passed", function() {
```

Comment:
I suggest pulling out "with a user that has been assigned a workspace" into a wrapping context (creating an extra level of nesting) to make this more straightforward.